### ISSUE TITLE
Tensorpack crash at first iteration

### ISSUE THREAD

**Issue by Rov67777** [tensorpack/tensorpack#813]
Hello,

I try to run the following code for densenets which uses tensorpack:
https://github.com/YixuanLi/densenet-tensorflow

I have made the following change to the code:
- remove "prefectData" and "imgaug.MapImage(lambda x: x - pp_mean)" from the augmentors because it did not run on windows
- change "SyncMultiGPUTrainer" to "SimpleTrainer" because I only have 1 GPU

My environnement:
- windows 7 pro
- Python 3.6.3
- Conda 4.4.7
- Tensorflow 1.8.0
- Tensorpack 0.8.6

python crashes at the first iteration of training, here is the log I get:

> block2/dense_layer.10/bn1/beta:0   [280]               280
> block2/dense_layer.10/conv1/W:0    [3, 3, 280, 12]   30240
> block2/dense_layer.11/bn1/gamma:0  [292]               292
> block2/dense_layer.11/bn1/beta:0   [292]               292
> block2/dense_layer.11/conv1/W:0    [3, 3, 292, 12]   31536
> block2/transition2/bn1/gamma:0     [304]               304
> block2/transition2/bn1/beta:0      [304]               304
> block2/transition2/conv1/W:0       [1, 1, 304, 304]  92416
> block3/dense_layer.0/bn1/gamma:0   [304]               304
> block3/dense_layer.0/bn1/beta:0    [304]               304
> block3/dense_layer.0/conv1/W:0     [3, 3, 304, 12]   32832
> block3/dense_layer.1/bn1/gamma:0   [316]               316
> block3/dense_layer.1/bn1/beta:0    [316]               316
> block3/dense_layer.1/conv1/W:0     [3, 3, 316, 12]   34128
> block3/dense_layer.2/bn1/gamma:0   [328]               328
> block3/dense_layer.2/bn1/beta:0    [328]               328
> block3/dense_layer.2/conv1/W:0     [3, 3, 328, 12]   35424
> block3/dense_layer.3/bn1/gamma:0   [340]               340
> block3/dense_layer.3/bn1/beta:0    [340]               340
> block3/dense_layer.3/conv1/W:0     [3, 3, 340, 12]   36720
> block3/dense_layer.4/bn1/gamma:0   [352]               352
> block3/dense_layer.4/bn1/beta:0    [352]               352
> block3/dense_layer.4/conv1/W:0     [3, 3, 352, 12]   38016
> block3/dense_layer.5/bn1/gamma:0   [364]               364
> block3/dense_layer.5/bn1/beta:0    [364]               364
> block3/dense_layer.5/conv1/W:0     [3, 3, 364, 12]   39312
> block3/dense_layer.6/bn1/gamma:0   [376]               376
> block3/dense_layer.6/bn1/beta:0    [376]               376
> block3/dense_layer.6/conv1/W:0     [3, 3, 376, 12]   40608
> block3/dense_layer.7/bn1/gamma:0   [388]               388
> block3/dense_layer.7/bn1/beta:0    [388]               388
> block3/dense_layer.7/conv1/W:0     [3, 3, 388, 12]   41904
> block3/dense_layer.8/bn1/gamma:0   [400]               400
> block3/dense_layer.8/bn1/beta:0    [400]               400
> block3/dense_layer.8/conv1/W:0     [3, 3, 400, 12]   43200
> block3/dense_layer.9/bn1/gamma:0   [412]               412
> block3/dense_layer.9/bn1/beta:0    [412]               412
> block3/dense_layer.9/conv1/W:0     [3, 3, 412, 12]   44496
> block3/dense_layer.10/bn1/gamma:0  [424]               424
> block3/dense_layer.10/bn1/beta:0   [424]               424
> block3/dense_layer.10/conv1/W:0    [3, 3, 424, 12]   45792
> block3/dense_layer.11/bn1/gamma:0  [436]               436
> block3/dense_layer.11/bn1/beta:0   [436]               436
> block3/dense_layer.11/conv1/W:0    [3, 3, 436, 12]   47088
> bnlast/gamma:0                     [448]               448
> bnlast/beta:0                      [448]               448
> linear/W:0                         [448, 10]          4480
> linear/b:0                         [10]                 10[36m
> Total #vars=119, #params=1019722, size=3.89MB[0m
> [32m[0706 11:39:19 @base.py:179][0m Setup callbacks graph ...
> [32m[0706 11:39:20 @predict.py:41][0m Building predictor tower 'InferenceTower
> ' on device /gpu:0 ...
> [32m[0706 11:39:20 @summary.py:38][0m Maintain moving average summary of 3 ten
> sors in collection MOVING_SUMMARY_OPS.
> [32m[0706 11:39:20 @summary.py:75][0m Summarizing collection 'summaries' of si
> ze 44.
> [32m[0706 11:39:20 @graph.py:91][0m Applying collection UPDATE_OPS of 78 ops.
> [32m[0706 11:39:26 @base.py:197][0m Creating the session ...
> 2018-07-06 11:39:26.870316: I T:\src\github\tensorflow\tensorflow\core\platform\
> cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow bi
> nary was not compiled to use: AVX2
> [32m[0706 11:39:27 @base.py:203][0m Initializing the session ...
> [32m[0706 11:39:27 @base.py:210][0m Graph Finalized.
> [32m[0706 11:39:30 @inference_runner.py:99][0m InferenceRunner will eval 157 i
> terations
> [32m[0706 11:39:31 @base.py:242][0m Start Epoch 1 ...
>   0%|                                                     |0/781[00:00<?,?it/s]

Thanks for your help

**Comment by Rov67777** [comment]
after investigation, the code seems to crash when calling "run_step" of base.py of the trainer

when executing the line  self.hooked_sess.run(self.train_op)
```

    def run_step(self):
        """
        Defines what to do in one iteration. The default is:
        ``self.hooked_sess.run(self.train_op)``.

        The behavior of each iteration can be changed by either setting ``trainer.train_op``,
        or overriding this method.
        """
        if not hasattr(self, 'train_op'):
            raise NotImplementedError(
                "Please either set `Trainer.train_op` or provide an implementation "
                "of Trainer.run_step()!")
        self.hooked_sess.run(self.train_op)
```

**Comment by ppwwyyxx** [comment]
The logs you posted does not contain any error. So the question is what makes you think it crashed?
The code runs well on linux. I recommend you to run some basic mnist examples  (in `examples/basics`) to rule out problems in your environment first.

Also, as mentioned in the issue template:
> 	+ Better to paste what you did instead of describing them.

**Comment by Rov67777** [comment]
Hey,

Thanks a lot for your answer. I pasted below the code with the changes.

To answer : "what makes you think it crashed"? Well code stops running at this point + output error message in windows : "Python has stopped working, A problem caused the program to stop working correctly"

I ran the MNIST examples in the basics and it worked correctly

`import numpy as np
import tensorflow as tf
import argparse
import os

from tensorpack import *
from tensorpack.tfutils.symbolic_functions import *
from tensorpack.tfutils.summary import *

"""
CIFAR10 DenseNet example. See: http://arxiv.org/abs/1608.06993
Code is developed based on Yuxin Wu's ResNet implementation: https://github.com/ppwwyyxx/tensorpack/tree/master/examples/ResNet
Results using DenseNet (L=40, K=12) on Cifar10 with data augmentation: ~5.77% test error.
Running time:
On one TITAN X GPU (CUDA 7.5 and cudnn 5.1), the code should run ~5iters/s on a batch size 64.
"""

BATCH_SIZE = 64

class Model(ModelDesc):
    def __init__(self, depth):
        super(Model, self).__init__()
        self.N = int((depth - 4)  / 3)
        self.growthRate =12

    def _get_inputs(self):
        return [InputDesc(tf.float32, [None, 32, 32, 3], 'input'),
                InputDesc(tf.int32, [None], 'label')
               ]

    def _build_graph(self, input_vars):
        image, label = input_vars
        image = image / 128.0 - 1

        def conv(name, l, channel, stride):
            return Conv2D(name, l, channel, 3, stride=stride,
                          nl=tf.identity, use_bias=False,
                          W_init=tf.random_normal_initializer(stddev=np.sqrt(2.0/9/channel)))
        def add_layer(name, l):
            shape = l.get_shape().as_list()
            in_channel = shape[3]
            with tf.variable_scope(name) as scope:
                c = BatchNorm('bn1', l)
                c = tf.nn.relu(c)
                c = conv('conv1', c, self.growthRate, 1)
                l = tf.concat([c, l], 3)
            return l

        def add_transition(name, l):
            shape = l.get_shape().as_list()
            in_channel = shape[3]
            with tf.variable_scope(name) as scope:
                l = BatchNorm('bn1', l)
                l = tf.nn.relu(l)
                l = Conv2D('conv1', l, in_channel, 1, stride=1, use_bias=False, nl=tf.nn.relu)
                l = AvgPooling('pool', l, 2)
            return l


        def dense_net(name):
            l = conv('conv0', image, 16, 1)
            with tf.variable_scope('block1') as scope:

                for i in range(self.N):
                    l = add_layer('dense_layer.{}'.format(i), l)
                l = add_transition('transition1', l)

            with tf.variable_scope('block2') as scope:

                for i in range(self.N):
                    l = add_layer('dense_layer.{}'.format(i), l)
                l = add_transition('transition2', l)

            with tf.variable_scope('block3') as scope:

                for i in range(self.N):
                    l = add_layer('dense_layer.{}'.format(i), l)
            l = BatchNorm('bnlast', l)
            l = tf.nn.relu(l)
            l = GlobalAvgPooling('gap', l)
            logits = FullyConnected('linear', l, out_dim=10, nl=tf.identity)

            return logits

        logits = dense_net("dense_net")

        prob = tf.nn.softmax(logits, name='output')

        cost = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=label)
        cost = tf.reduce_mean(cost, name='cross_entropy_loss')

        wrong = prediction_incorrect(logits, label)
        # monitor training error
        add_moving_summary(tf.reduce_mean(wrong, name='train_error'))

        # weight decay on all W
        wd_cost = tf.multiply(1e-4, regularize_cost('.*/W', tf.nn.l2_loss), name='wd_cost')
        add_moving_summary(cost, wd_cost)

        add_param_summary(('.*/W', ['histogram']))   # monitor W
        self.cost = tf.add_n([cost, wd_cost], name='cost')

    def _get_optimizer(self):
        lr = tf.get_variable('learning_rate', initializer=0.1, trainable=False)
        tf.summary.scalar('learning_rate', lr)
        return tf.train.MomentumOptimizer(lr, 0.9, use_nesterov=True)


def get_data(train_or_test):
    isTrain = train_or_test == 'train'
    ds = dataset.Cifar10(train_or_test)
    pp_mean = ds.get_per_pixel_mean()
    if isTrain:
        augmentors = [
            imgaug.CenterPaste((40, 40)),
            imgaug.RandomCrop((32, 32)),
            imgaug.Flip(horiz=True),
            #imgaug.Brightness(20),
            #imgaug.Contrast((0.6,1.4)),
            #imgaug.MapImage(lambda x: x - pp_mean),
        ]
    else:
        augmentors = [
            #imgaug.MapImage(lambda x: x - pp_mean)
        ]
    ds = AugmentImageComponent(ds, augmentors)
    ds = BatchData(ds, BATCH_SIZE, remainder=not isTrain)
    #if isTrain:
        #ds = PrefetchData(ds, 3, 2)
    return ds

def get_config():
    log_dir = 'train_log/cifar10-single-fisrt%s-second%s-max%s' % (str(args.drop_1), str(args.drop_2), str(args.max_epoch))
    logger.set_logger_dir(log_dir, action='n')

    # prepare dataset
    dataset_train = get_data('train')
    steps_per_epoch = dataset_train.size()
    dataset_test = get_data('test')

    return TrainConfig(
        dataflow=dataset_train,
        callbacks=[
            ModelSaver(),
            InferenceRunner(dataset_test,
                [ScalarStats('cost'), ClassificationError()]),
            ScheduledHyperParamSetter('learning_rate',
                                      [(1, 0.1), (args.drop_1, 0.01), (args.drop_2, 0.001)])
        ],
        model=Model(depth=args.depth),
        steps_per_epoch=steps_per_epoch,
        max_epoch=args.max_epoch,
    )

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--gpu', help='comma separated list of GPU(s) to use.') # nargs='*' in multi mode
    parser.add_argument('--load', help='load model')
    parser.add_argument('--drop_1',default=150, help='Epoch to drop learning rate to 0.01.') # nargs='*' in multi mode
    parser.add_argument('--drop_2',default=225,help='Epoch to drop learning rate to 0.001')
    parser.add_argument('--depth',default=40, help='The depth of densenet')
    parser.add_argument('--max_epoch',default=300,help='max epoch')
    args = parser.parse_args()

    if args.gpu:
        os.environ['CUDA_VISIBLE_DEVICES'] = args.gpu

    config = get_config()
    if args.load:
        config.session_init = SaverRestore(args.load)
    
    nr_tower = 0
    if args.gpu:
        nr_tower = len(args.gpu.split(','))
    
    # SyncMultiGPUTrainer(config).train()
    launch_train_with_config(config, SimpleTrainer())`

**Comment by ppwwyyxx** [comment]
Your modification still runs well on Linux. I believe this is an environment problem (software incompatibilities among each other, etc). Sorry but I cannot help.

**Comment by Rov67777** [comment]
Hey, thanks for your answers,

- About it being an OOM : I tried reducing the depth / number of parameters to very low amounts and it still not run so I don't think this is the problem
- About testing other official code: actually the cifar-convnet (from basic) does not run either with the same crash (crash at the first iteration).

**Comment by ppwwyyxx** [comment]
Probably an issue with opencv. When tensorflow first released it has some bad interaction with certain build of opencv, on Linux. Maybe the problem still exists on windows.
For example, if your opencv is built with cuda support or opencl support, it's very likely to cause crash when used together with tensorflow. There are other possible causes as well.

**Comment by Rov67777** [comment]
Hey,

It is indeed a problem with opencv. More precisely when using the following function:
`ret = cv2.flip(img, self.code)
`            
in line 46 of the image augmentation package: https://github.com/tensorpack/tensorpack/blob/master/tensorpack/dataflow/imgaug/misc.py

I have changed the code to the numpy version of flip which works correctly. Altough, I guess I might encounter problem performance wise by using numpy instead of opencv to compute the augmentation ?

`ret = np.flip(img, self.code)
`

Well, if not perfect, at least it works. If somebody has any idea how i could try to debug the problem with the opencv call to flip it would be great. Otherwise the issue can be closed.

Thanks all for the help

**Comment by ppwwyyxx** [comment]
I'm closing this now since it's not a tensorpack problem. Maybe `cv2.getBuildInformation()` can tell something, e.g., you can check whether CUDA or opencl is enabled in `cv2.getBuildInformation()`.
